src/nnet3/discriminative-training.cc

```diff
@@ -0,0 +1,602 @@
+  // This can be a size zero vector e.g. for 'chain' model
+  const CuVectorBase<BaseFloat> &log_priors_;
+
+  const DiscriminativeSupervision &supervision_;
+  
+  // The neural net output.
+  const CuMatrixBase<BaseFloat> &nnet_output_;
+
+  // Training stats including accumulated objective function, gradient
+  // and total weight. Optionally the nnet_output and gradients per pdf can be
+  // accumulated for debugging purposes.
+  DiscriminativeTrainingStats *stats_;
+
+  // l2 regularization constant on the 'chain' output; the actual term added to
+  // the objf will be -0.5 times this constant times the squared l2 norm.
+  // (squared so it's additive across the dimensions).  e.g. try 0.0005.
+  BaseFloat *l2_term_;
+
+  // If non-NULL, derivative w.r.t. to nnet_output is written here.
+  CuMatrixBase<BaseFloat> *nnet_output_deriv_;
+   
+  // If non-NULL, then the xent objective derivative 
+  // (which equals a posterior from the numerator forward-backward, scaled by
+  // the supervision weight) is written to here.  
+  // This will be used in the cross-entropy regularization code.  
+  CuMatrixBase<BaseFloat> *xent_output_deriv_;
+
+  // Denominator lattice. 
+  Lattice den_lat_;
+
+  // List of silence phones. Useful to treat silence phones
+  // differently in computing SMBR / MPFE objectives.
+  std::vector<int32> silence_phones_;
+
+  // The function that actually computes the objective and gradients
+  double ComputeObjfAndDeriv(Posterior *post, Posterior *xent_post);
+
+  static inline Int32Pair MakePair(int32 first, int32 second) {
+    Int32Pair ans;
+    ans.first = first;
+    ans.second = second;
+    return ans;
+  }
+};
+
+DiscriminativeComputation::DiscriminativeComputation(
+                            const DiscriminativeTrainingOptions &opts,
+                            const TransitionModel &tmodel,
+                            const CuVectorBase<BaseFloat> &log_priors,
+                            const DiscriminativeSupervision &supervision,
+                            const CuMatrixBase<BaseFloat> &nnet_output,
+                            DiscriminativeTrainingStats *stats,
+                            BaseFloat *l2_term,
+                            CuMatrixBase<BaseFloat> *nnet_output_deriv,
+                            CuMatrixBase<BaseFloat> *xent_output_deriv)
+  : opts_(opts), tmodel_(tmodel), log_priors_(log_priors), 
+  supervision_(supervision), nnet_output_(nnet_output),
+  stats_(stats), 
+  l2_term_(l2_term),
+  nnet_output_deriv_(nnet_output_deriv), 
+  xent_output_deriv_(xent_output_deriv) {
+  
+  den_lat_ = supervision.den_lat;
+  TopSort(&den_lat_);
+  
+  if (!SplitStringToIntegers(opts_.silence_phones_str, ":", false,
+                             &silence_phones_)) {
+    KALDI_ERR << "Bad value for --silence-phones option: "
+              << opts_.silence_phones_str;
+  }
+}
+
+void DiscriminativeComputation::Compute() {
+  if (opts_.criterion == "mmi" && opts_.boost != 0.0) {
```

Comment:
there is a 'soft' 40-line limit on functions.
you have blown *way* past the limit with this function.